### ISSUE TITLE
setInput(_:) not changing the value associated with a TextField

### ISSUE THREAD

**Issue by grantneufeld** [nalexn/ViewInspector#182]
**Update: I’ve narrowed the problem down to `setInput` not actually setting the value of the text in the TextField. See first comment below for details.**

---

I have a SwiftUI view that conditionally shows a `Text` view (an error message), or an `EmptyView`, based on an observed var attached to a `TextField` (in the code below: `data.name`).

When I use ViewInspector’s `setInput(_:)` method on the TextField, the `onChange(of:,perform:)` closure attached to the `TextField` does not get triggered.

Running the code in Simulator, editing the text in the TextField does indeed trigger the onChange closure (hiding the Text view, if it is showing). So, the code itself is working, but I haven’t figured out how to get ViewInspector to trigger that “edited” effect on the TextField.

### View:

```swift
struct NewUserFormView: View {
    @ObservedObject var data: UserFormData
    var body: some View {
        Form {
            TextField("Name", text: $data.name)
                // the onChange closure
                .onChange(of: data.name) { _ in
                    data.nameInvalidMessage = ""
                }
            // the conditionally displayed error message
            if !data.nameInvalidMessage.isEmpty {
                Text(data.nameInvalidMessage)
            } else {
                EmptyView()
            }
            // ... (more form elements not relevant to this discussion)
        }
    }
}

class UserFormData: ObservableObject {
    @Published var name = ""
    @Published var nameInvalidMessage = ""
}
```

### Test:

```swift
@testable import MyApp
import ViewInspector
import XCTest

class NewUserFormViewTests: XCTestCase {
    func testExample() throws {
        userData = UserFormData()
        sut = NewUserFormView(data: userData)
        view = try sut.inspect().view(NewUserFormView.self)
        let errorMessage = "Validation failed."
        // Verify that the validation error message is not being shown:
        XCTAssertThrowsError(try view.find(text: errorMessage)) // 🟢 passes
        // cause a validation error
        userData.nameInvalidMessage = errorMessage
        // Verify that the validation error message is being shown:
        XCTAssertNoThrow(try view.find(text: errorMessage)) // 🟢 passes
        // edit the field
        try view.findTextFieldWith(name: "Name").setInput("Edited")
        // Verify that the validation error message is not being shown:
        XCTAssertThrowsError(try view.find(text: errorMessage)) // 🛑 This fails
    }
}

extension NewUserFormView: Inspectable {}
```

### Test Helpers

```swift
extension InspectableView {
    func findTextFieldWith(name: String) throws -> InspectableView<ViewType.TextField> {
        try find(
            ViewType.TextField.self,
            where: { view in
                let label = try view.labelView().text().string()
                return label == name
            }
        )
    }
}
```

**Comment by grantneufeld** [comment]
After some more digging (manually stepping through the code in the debugger), I’ve found that `setInput` isn’t actually setting the input on the given `TextField`.

```swift
// From: Sources/ViewInspector/SwiftUI/TextField.swift

public extension InspectableView where View == ViewType.TextField {
    // ...
    func setInput(_ value: String) throws {
        try guardIsResponsive()
        try inputBinding().wrappedValue = value
    }
    // ...
}
```

It’s not throwing an error at either step (it passes all the checks in `guardIsResponsive()` and `inputBinding()`), so I’ve been unable to figure out why it’s failing to assign the String value to the TextField.

**Comment by boyceEstes** [comment]
I'm having problems with this as well. Created a minimum sample to show the issue.

```
    func test_textfield_inputValue_deliversExpectedValue() throws {
        // given
        let expectedInput = "Success!"
        
        @State var input = ""
        let sut = TextField("Input", text: $input)
        
        let textField = try sut
            .inspect()
            .find(ViewType.TextField.self)
            
        // when
        try textField.setInput(expectedInput)
        
        // then
        let newTextFieldValue = try textField.input()
        XCTAssertEqual(newTextFieldValue, expectedInput)

// OUTPUT: test_textfield_inputValue_deliversExpectedValue(): XCTAssertEqual failed: ("") is not equal to ("Success!")
    }
```

**Comment by boyceEstes** [comment]
Also having a problem with this. I've created a minimum test to showcase the issue.

```
    func test_textfield_inputValue_deliversExpectedValue() throws {
        // given
        let expectedInput = "Success!"
        
        @State var input = ""
        let sut = TextField("Any", text: $input)
        
        let textField = try sut
            .inspect()
            .find(ViewType.TextField.self)
            
        // when
        try textField.setInput(expectedInput)
        
        // then
        let newTextFieldValue = try textField.input()
        XCTAssertEqual(newTextFieldValue, expectedInput)
    }
```
Error:  `test_textfield_inputValue_deliversExpectedValue(): XCTAssertEqual failed: ("") is not equal to ("Success!")`

**Comment by nalexn** [comment]
My apologies for the late response. The problem lies in the nuances of SwiftUI state management, the `.onChange` from the 1st example and `@State` from the 2nd require the view to be hosted on screen for the state management to propagate to the view.

Fixed 1st test:

```swift
func testExample() throws {
    guard #available(iOS 14.0, macOS 10.15, tvOS 14.0, *)
    else { throw XCTSkip() }
    let userData = UserFormData()
    let sut = NewUserFormView(data: userData)
    let errorMessage = "Validation failed."
    let exp1 = sut.inspection.inspect { view in
        // Verify that the validation error message is not being shown:
        XCTAssertThrowsError(try view.find(text: errorMessage)) // 🟢 passes
        // cause a validation error
        userData.nameInvalidMessage = errorMessage
        // Verify that the validation error message is being shown:
        XCTAssertNoThrow(try view.find(text: errorMessage)) // 🟢 passes
        // edit the field
        try view.findTextFieldWith(name: "Name").setInput("Edited")
    }
    let exp2 = sut.inspection.inspect(after: 0.1) { view in
        // Verify that the validation error message is not being shown:
        XCTAssertThrowsError(try view.find(text: errorMessage)) // 🛑 This fails
    }
    ViewHosting.host(view: sut)
    wait(for: [exp1, exp2], timeout: 0.3)
}

struct NewUserFormView: View {
    @ObservedObject var data: UserFormData
    let inspection = Inspection<Self>()
    
    var body: some View {
        Form {
            ...
        }
        .onReceive(inspection.notice) { self.inspection.visit(self, $0) }
    }
}
```

Fixed 2nd test (without hosting, just by replacing `@State` with `Binding`):

```swift
func test_textfield_inputValue_deliversExpectedValue() throws {
    // given
    let expectedInput = "Success!"
    
    let input = Binding(wrappedValue: "")
    let sut = TextField("Any", text: input)
    
    let textField = try sut
        .inspect()
        .find(ViewType.TextField.self)
        
    // when
    try textField.setInput(expectedInput)
    
    // then
    let newTextFieldValue = try textField.input()
    XCTAssertEqual(newTextFieldValue, expectedInput)
}
```
